src/glang/lib/std/rb_tree.c3

```diff
@@ -0,0 +1,273 @@
+
+        // Case 4: u has no parent.
+        if !u.parent:has_value() {
+            // u is the root node. Colour it black.
+            u.colour = BLACK();
+            return;
+        }
+
+        let v : RB_Node<K, V>& = &u.parent:data();
+
+        // Case 1: u has a black parent.
+        if v.colour == BLACK() {
+            // u is red, so both of its children must be black. Given that its
+            // parent is also black, we can just leave u red.
+            return;
+        }
+
+        // u has a red parent. Property 2 of red-black trees states that the
+        // root node is black, therefore v also has a parent.
+        let w : RB_Node<K, V>& = &v.parent:data();
+
+        // Case 2: u has a red parent and a red uncle.
+        if w.left:is_red() and w.right:is_red() {
+            w.left:colour_black();
+            w.right:colour_black();
+
+            // Push the issue up the tree.
+            w.colour = RED();
+            u_ptr = {&w};
+            // TODO continue;
+        }
+        else {
+            // Case 3: u has a red parent and a black uncle.
+            const u_side : bool = which_child_am_i(&u);
+            const uncle_side : bool = !which_child_am_i(&v);
+
+            if u_side != uncle_side {
+                // Case 3a: u is away from its uncle.
+                // Rotate, then leave u red and colour w, its new sibling, red.
+                // Their new parent should thus be black.
+                self:rotate(&w, !u_side);
+                v.colour = BLACK();
+                w.colour = RED();
+            }
+            else {
+                // Case 3b: u is towards its uncle.
+                // Now u becomes the parent of v and w. Colour it black, leave v
+                // red, and make w red.
+                self:rotate(&v, !u_side);
+                self:rotate(&w, u_side);
+                u.colour = BLACK();
+                w.colour = RED();
+            }
+
+            return;
+        }
+    }
+}
+
+function [K, V] insert : (
+    self : RB_Tree<K, V>&, key : K, value : V
+) -> void = {
+    let p_opt : Optional<RB_Node<K, V>&> = make_optional<RB_Node<K, V>&>();
+    let u_ptr : Ptr<Optional<RB_Node<K, V>&>> = {&self.root};
+
+    while u_ptr:data():has_value() {
+        let u : RB_Node<K, V>& = &u_ptr:data():data();
+
+        if key == u.key {
+            // Key already exists. We can just replace the value and return
+            // early, as the red-black tree is already valid.
+            u.value = value;
+            return;
+        }
+
+        // Keep track of the parent node.
+        p_opt:store(&u_ptr:data():data());
+
+        u_ptr = {&u:get_child(left_if(key < u.key))};
+    }
+
+    u_ptr:data():store(&self:new<RB_Node<K, V>>(key, value, p_opt));
+
+    if p_opt:has_value() {
+        let p : RB_Node<K, V>& = &p_opt:data();
+        p:get_child(left_if(key < p.key)) = u_ptr:data();
+    }
+
+    self:fix_after_insert(&u_ptr:data():data());
+}
```

Comment:
I think this should be part of the `map`s interface not necessarily a general purpose red-black tree